Initial

aljones15 opened the pull request:
Web based service for [bedrock-payment-http](https://github.com/digitalbazaar/bedrock-payment-http/pull/1)

aljones15 committed:
Add initial files.

aljones15 committed:
Use bedrock-payment module update peerDeps.

aljones15 committed:
Add paypal btn and linter for vue.

aljones15 committed:
Update author.

aljones15 committed:
Add license.

aljones15 committed:
Add README.

aljones15 committed:
Add Payment class to web.

aljones15 committed:
Start Payment with status PENDING.

aljones15 committed:
Update CHANGELOG.

aljones15 committed:
Update docs on Payment.

aljones15 committed:
Move PaymentService into it's own class.

aljones15 committed:
Add the ability to disable cards.

aljones15 committed:
Add PaymentService to Change Log.

mattcollier commented on lib/PaymentService.js:
```diff
@@ -0,0 +1,82 @@
+export default class PaymentService {
+  constructor({baseURL = '/payment', httpsAgent = axios} = {}) {
+    this.baseURL = baseURL;
+    this.httpsAgent = httpsAgent;
```
I'm concerned that the use of `httpsAgent` here does not seem consistent with how that term is used elsewhere, for instance: https://github.com/digitalbazaar/web-ledger-client/blob/master/lib/WebLedgerClient.js#L31-L32

Also, seems to me that services need to define some good defaults for axios which would include setting a timeout value instead of the silly default of no timeout.

aljones15 replied:
I think it might be better to make this module more light weight and remove axios entirely and just throw if no httpsAgent is there. That said the name for httpsAgent might be confusing with back end code, but I thought making the module used to make the actual api call configurable might be better in the future.

mattcollier replied:
Well, this points to a larger issue in the front end, which is that if services like this do not abstract away the interface and errors produced by a particular xhr request library, then then there is really no ability to swap out one module for another. I know that you made an attempt to address the error handling in the past, but that is still an unsolved problem today.

aljones15 replied:
so what you are saying is that as long as we can not get consistent error messages across multiple modules we have a problem. We could decide on any library that implements the fetch api: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

There are plenty of existing npm projects that implement it.

mattcollier commented on package.json:
```diff
@@ -0,0 +1,42 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/digitalbazaar/bedrock-web-payment.git"
```
For the publishing scripts to work properly, this needs to be an HTTPS url, removing the default `git+`

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/0618c257af143d5a0adf0ea68f7f4f84c59f9358

and thanks for this one.

mattcollier commented on package.json:
```diff
@@ -0,0 +1,42 @@
+  "homepage": "https://github.com/digitalbazaar/bedrock-web-payment#readme",
+  "peerDependencies": {
+    "bedrock-payment": "^0.0.1",
+    "axios": "^0.19.0"
```
axios looks like a regular dependency to me.

aljones15 replied:
Removed Axios entirely here: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/0618c257af143d5a0adf0ea68f7f4f84c59f9358

mattcollier replied:
I don't think that's what we want.  IMO, consumers of services modules should not have to concern themselves with specifying an HTTP module to use.  See my other comment on this topic as well.

aljones15 replied:
Bundle size is what I'm concerned about. Axios is not a huge library, but if we didn't care about features or backwards compatibility we could use `fetch` directly or newer modules like `mappersmith`. On the backend including a node `httpsAgent` provides little overhead for the customer as they are already using node.

aljones15 replied:
Added axios as a dependency: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/86b77edc8ddf6279e6be7ca30b44a6dae20d1bfc

mattcollier commented on package.json:
```diff
@@ -0,0 +1,42 @@
+  },
+  "homepage": "https://github.com/digitalbazaar/bedrock-web-payment#readme",
+  "peerDependencies": {
+    "bedrock-payment": "^0.0.1",
```
Historically, front end modules do not specify back-end dependencies or peer deps, for example: https://github.com/digitalbazaar/bedrock-web-account/blob/master/package.json

aljones15 replied:
It might be a good idea to question this convention as I think the warning about not having the peerDependency `bedrock-payment-http` might be a good idea as it would inform a developer that you installed a front end module, but forgot the back end module. However this is `bedrock-payment` which is no longer needed.

mattcollier replied:
There is no requirement that the front-end and back-end live on the same server.  In fact we look forward to the day when the front-end bundle can live on a CDN.

aljones15 replied:
That is a valid point. Removed all peerDependencies here: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/11f3885c9cfb225ebbecb915db0a33afd46a99ff

mattcollier commented on lib/Payment.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {string} [options.currency = 'USD'] - The currency of the payment.
+ * @param {Array<string|object>} options.orders - The productIds of the products
+ *   being purchased.
+ * @param {string} [options.stats = PaymentStatus.PENDING] - The payment status.
```
does `PaymentStatus.PENDING` correspond to something? I see the status in the class is set to the string 'PENDING'.

aljones15 replied:
It does correspond to a set of statuses on the back end: https://github.com/digitalbazaar/bedrock-payment/blob/initial/lib/constants.js

I did change the comment to use 'PENDING' as the default value, but I actually think in this case referencing PaymentStatus might be more information for a developer in the future.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/0618c257af143d5a0adf0ea68f7f4f84c59f9358

mattcollier reviewed (changes requested):
Nicely done!  Congratulations on your first bedrock module! :)

A few comments/concerns.

aljones15 committed:
Remove +git from package.json repo name.

  - Remove axios peerDependency.
  - Remove PaymentStatus from class type.

aljones15 committed:
Remove peerDepencies.

aljones15 committed:
Readd axios to project.

aljones15 committed:
Use axios.

mattcollier reviewed (approved):
TY!

gannan08 commented on components/PayPalButton.vue:
```diff
@@ -0,0 +1,177 @@
+    },
+    disableCard: {
+      type: Array,
+      default: () => ['amex']
```
Why should this component disable `amex` cards by default? Suppose, Company A sets up its PayPal Account in a way that does not support AMEX cards and Company B does. This design decision now forces Company B to pass an empty array in order to disable no cards. This seems awkward. It would make sense to me that this array is empty and the user of the API decides what cards should be disabled based on their own use case or business problem(s).

aljones15 replied:
I actually agree this with this. I put `amex` in there, because my PayPal account is not a pro-account and can not accept it yet. Addressed here: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/ad2507115de81dcf96f940e9ce7172f59760240a

gannan08 commented on lib/Payment.js:
```diff
@@ -0,0 +1,47 @@
+    error = null
+  }) {
+    this.id = id;
+    // https://developer.paypal.com/docs/api/reference/currency-codes/
```
Why the comment to PayPal docs for currency codes above the `amount` field? It is my understanding that this Payment class is agnostic to any Payment Gateway. If true then why refer to PayPal docs. If it is the case we need to refer to currency codes then there should be some ISO Standard or some standard developed at an SDO that specifies such codes.

aljones15 replied:
I removed this, but I do feel that despite being gateway specific it at least did give the developer some information on what a currency code should look like: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/ad2507115de81dcf96f940e9ce7172f59760240a

gannan08 commented on lib/PaymentService.js:
```diff
@@ -0,0 +1,79 @@
+ * @returns {PaymentService} The PaymentService.
+ */
+export default class PaymentService {
+  constructor({baseURL = '/payment'} = {}) {
```
Please use the `urls` pattern for defining urls in a service.

See https://github.com/digitalbazaar/bedrock-web-account/blob/master/AccountService.js#L17-L23

aljones15 replied:
Not really needed here at the moment, but maybe we can expand it in the future: https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/ad2507115de81dcf96f940e9ce7172f59760240a

gannan08 replied:
Why do you believe it's not needed here at the moment? We want to consistently define our `urls` across all of our services we make in bedrock modules.

aljones15 replied:
ok did not know this was a pattern (despite the fact that I have written at least one of these before). TokenService also uses it: https://github.com/digitalbazaar/bedrock-web-authn-token/blob/master/TokenService.js

I just through it was not needed because we only have one URL. I'm assuming we will have more in the future?

gannan08 replied:
The subtlety here is the assumption that we only have one URL **right now**. That may not be true in the future. On top of that, this structure breaks away from the common pattern we want to have with web services. Personally, it is paramount to maintain the consistency of our libraries as much as possible. It reduces the cognitive load when picking up something new. If built properly, the library should feel familiar to a bedrock developer.

gannan08 commented on lib/PaymentService.js:
```diff
@@ -0,0 +1,79 @@
+   *
+   * @returns {Promise<object>} The meta data for the Payment Button.
+   */
+  async getMeta({service = 'paypal'} = {}) {
```
This API says we're getting meta data but the created url looks like we're getting credentials. Is it the case that you're making an HTTP Request to receive secrets/credentials from the backend to complete charges on the client?

aljones15 replied:
I changed it to `getCredentials`. Sorry about this it was getMeta before renaming it. I'm assuming credentials is the right name for payment gateway oauth tokens, api keys, etc.

https://github.com/digitalbazaar/bedrock-web-payment/pull/1/commits/ad2507115de81dcf96f940e9ce7172f59760240a

aljones15 replied:
@gannan08 I addressed all of your concerns if you are happy with the PR feel free to rebase it as @mattcollier has approved it.

gannan08 replied:
> I changed it to `getCredentials`. Sorry about this it was getMeta before renaming it. I'm assuming credentials is the right name for payment gateway oauth tokens, api keys, etc.
> 
> [ad25071](https://github.com/digitalbazaar/bedrock-web-payment/commit/ad2507115de81dcf96f940e9ce7172f59760240a)

Thank you for renaming it. What is in the response?

aljones15 replied:
for the accelerator `paypal-client-id` 

```
{
  service: 'paypal',
  paypal-client-id: 'client-id'
}

gannan08 replied:
Are you saying this is the specific response for just the accelerator? Shouldn't this be the same response regardless of what product this gets built into? Additionally, we should not be using kebab case in our APIs.

aljones15 replied:
@gannan08 https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/0eec0778ef91dcd8dcdb499321376e39d7f3ecc4

aljones15 committed:
Default disabled cards to [].

  - Remove PayPal specific comment.
  - Use the AccountService urls object.
  - Change getMeta to getCredentials.
